vere/main.c

```diff
@@ -546,6 +588,59 @@ main(c3_i   argc,
     return 1;
   }
 
+  /*
+    In daemon mode, run the urbit as a background process, but don't
+    exit from the parent process until the ship is finished booting.
+
+    We use a pipe to communicate between the child and the parent. The
+    parent waits for the child to write something to the pipe and
+    then exits. If the pipe is closed with nothing written to it, get
+    the exit status from the child process and also exit with that status.
+
+    We want the child to write to the pipe once it's booted, so we put
+    `_on_boot_completed_cb` into `u3_Host.onboot`, which is NULL in
+    non-daemon mode. That gets called once the `chis` service is
+    available.
+
+    In both processes, we are good fork() citizens, and close all unused
+    file descriptors. Closing `pipefd[1]` in the parent process is
+    especially important, since the pipe needs to be closed if the child
+    process dies. When the pipe is closed, the read fails, and that's
+    how we know that something went wrong.
+
+    There are some edge cases around `WEXITSTATUS` that are not handled
+    here, but I don't think it matters.
+  */
+  if ( c3y == u3_Host.ops_u.dem ) {
```

Comment:
I think it might be nicer if the contents of this block were in a separate function.